README.md
<div align="center">
<h1>Paper Onboarding</h1>

[![npm](https://badgen.net/npm/v/@gorhom/paper-onboarding)](https://www.npmjs.com/package/@gorhom/paper-onboarding) [![npm](https://badgen.net/npm/license/@gorhom/paper-onboarding)](https://www.npmjs.com/package/@gorhom/paper-onboarding) [![npm](https://badgen.net/npm/types/@gorhom/paper-onboarding)](https://www.npmjs.com/package/@gorhom/paper-onboarding)

<img src="./preview.gif">

Paper Onboarding is a material design UI slider for `React Native` inspired by [Ramotion Paper Onboarding](https://github.com/Ramotion/paper-onboarding).

</div>

---

## Installation

```sh
yarn add @gorhom/paper-onboarding
# or
npm install @gorhom/paper-onboarding
```

> Also, you need to install [react-native-reanimated](https://github.com/software-mansion/react-native-reanimated), [react-native-gesture-handler](https://github.com/software-mansion/react-native-gesture-handler) & [react-native-svg](https://github.com/react-native-community/react-native-svg), and follow their installation instructions.

## Usage

```tsx
import PaperOnboarding, {PaperOnboardingItemType} from "@gorhom/paper-onboarding";

const data: PaperOnboardingItemType[] = [
  {
    title: 'Hotels',
    description: 'All hotels and hostels are sorted by hospitality rating',
    backgroundColor: '#698FB8',
    image: /* IMAGE COMPONENT */,
    icon: /* ICON COMPONENT */,
    content: /* CUSTOM COMPONENT */,
  },
  {
    title: 'Banks',
    description: 'We carefully verify all banks before add them into the app',
    backgroundColor: '#6CB2B8',
    image: /* IMAGE COMPONENT */,
    icon: /* ICON COMPONENT */,
    content: /* CUSTOM COMPONENT */,
  },
  {
    title: 'Stores',
    description: 'All local stores are categorized for your convenience',
    backgroundColor: '#9D8FBF',
    image: /* IMAGE COMPONENT */,
    icon: /* ICON COMPONENT */,
    content: /* CUSTOM COMPONENT */,
  },
];

const Screen = () => {
  const handleOnClosePress = () => console.log('navigate to other screen')
  return (
    <PaperOnboarding
      data={data}
      onCloseButtonPress={handleOnClosePress}
    />
  )
}
```

## Props

| name                     | description                                                          | required | type                                                       | default                                    |
| ------------------------ | -------------------------------------------------------------------- | -------- | ---------------------------------------------------------- | ------------------------------------------ |
| data                     | Array of pages/slides to present.                                    | YES      | Array<[PaperOnboardingItemType](#paperonboardingitemtype)> |                                            |
| safeInsets               | Safe area insets usually come from `react-native-safe-area-context`. | NO       | Insets                                                     | {top: 50, bottom: 50, left: 50, right: 50} |
| direction                | Pan gesture direction.                                               | NO       | 'horizontal' \| 'vertical'                                 | horizontal                                 |
| indicatorSize            | Indicator size (width and height).                                   | NO       | number                                                     | 40                                         |
| indicatorBackgroundColor | Indicator background color.                                          | NO       | string                                                     | white                                      |
| indicatorBorderColor     | Indicator border color.                                              | NO       | string                                                     | white                                      |
| titleStyle               | Text style to override `all` page/slide title style.                 | NO       | StyleProp<TextStyle>                                       |                                            |
| descriptionStyle         | Text style to override `all` page/slide description style.           | NO       | StyleProp<TextStyle>                                       |                                            |
| closeButton              | Custom component to be used instead of the default close button.     | NO       | (() => React.ReactNode) \| React.ReactNode                 |                                            |
| closeButtonText          | Close button text.                                                   | NO       | string                                                     | close                                      |
| closeButtonTextStyle     | Close button text style.                                             | NO       | StyleProp<TextStyle>                                       |                                            |
| onCloseButtonPress       | Callback on close button pressed.                                    | NO       | () => void                                                 |                                            |
| onIndexChange            | Callback when index change.                                          | NO       | () => void                                                 |                                            |

#### PaperOnboardingItemType

| name             | description                                                                                   | required | type                                                                                    |
| ---------------- | --------------------------------------------------------------------------------------------- | -------- | --------------------------------------------------------------------------------------- |
| content          | Slide/page content, this will replace default content.                                        | NO       | ((props: [PageContentProps](./src/types.ts#L87)) => React.ReactNode) \| React.ReactNode |
| image            | Image cover.                                                                                  | NO       | (() => React.ReactNode) \| React.ReactNode                                              |
| icon             | Indicator icon.                                                                               | NO       | (() => React.ReactNode) \| React.ReactNode                                              |
| backgroundColor  | Background color.                                                                             | YES      | string                                                                                  |
| title            | Title text.                                                                                   | NO       | string                                                                                  |
| description      | Description text.                                                                             | NO       | string                                                                                  |
| titleStyle       | Text style to override page/slide title default style.                                        | NO       | StyleProp<TextStyle>                                                                    |
| descriptionStyle | Text style to override page/slide description default style.                                  | NO       | StyleProp<TextStyle>                                                                    |
| showCloseButton  | Show close button when page/slide is active, _note: last page will always show close button._ | NO       | boolean                                                                                 |

## Methods

| name     | description                | type       |
| -------- | -------------------------- | ---------- |
| next     | Navigate to next page.     | () => void |
| previous | Navigate to previous page. | () => void |

## Built With ❤️

- [react-native-gesture-handler](https://github.com/software-mansion/react-native-gesture-handler)
- [react-native-reanimated](https://github.com/software-mansion/react-native-reanimated)
- [react-native-redash](https://github.com/wcandillon/react-native-redash)
- [react-native-svg](https://github.com/react-native-community/react-native-svg)
- [@react-native-community/bob](https://github.com/react-native-community/bob)

## Author

- [Mo Gorhom](https://twitter.com/gorhom)

## License

MIT

<div align="center">

Liked the library? 😇

<a href="https://www.buymeacoffee.com/gorhom" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/default-red.png" alt="Buy Me A Coffee" height="50" ></a>

</div>

---

<p align="center">
<a href="https://gorhom.dev" target="_blank"><img alt="Mo Gorhom" src="./logo.png"></a>
</p>


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v1.2.0](https://github.com/gorhom/react-native-paper-onboarding/compare/v1.1.0...v1.2.0)

- feat: added next & previous methods [`#34`](https://github.com/gorhom/react-native-paper-onboarding/pull/34)
- feat: added RTL support (#32) by @aleppos [`43a61e4`](https://github.com/gorhom/react-native-paper-onboarding/commit/43a61e410b3599e3666f79a758c6cab5bb8ae738)
- chore: updated build script [`1a9c307`](https://github.com/gorhom/react-native-paper-onboarding/commit/1a9c307e63dabc503be3787d0e11619a78d46601)

#### [v1.1.0](https://github.com/gorhom/react-native-paper-onboarding/compare/v1.0.0...v1.1.0)

> 16 June 2021

- docs: updated readme.md [`#25`](https://github.com/gorhom/react-native-paper-onboarding/pull/25)
- chore: updated dependencies [`57b497f`](https://github.com/gorhom/react-native-paper-onboarding/commit/57b497fe2b64803a14c39ea74680451883003230)
- feat: added Reanimated v2 compatibility [`0f1eacd`](https://github.com/gorhom/react-native-paper-onboarding/commit/0f1eacdeba103d25bf51d4d859247af12869fb74)

### [v1.0.0](https://github.com/gorhom/react-native-paper-onboarding/compare/v0.5.0...v1.0.0)

> 14 June 2020

- chore: update branding [`#24`](https://github.com/gorhom/react-native-paper-onboarding/pull/24)
- feat: optimise performance and internal refactor [`#23`](https://github.com/gorhom/react-native-paper-onboarding/pull/23)
- refactor: indicators and close button [`#22`](https://github.com/gorhom/react-native-paper-onboarding/pull/22)
- feat: export page content props [`#21`](https://github.com/gorhom/react-native-paper-onboarding/pull/21)
- feat: enable dismissible pages/slides [`#20`](https://github.com/gorhom/react-native-paper-onboarding/pull/20)
- refactor: project structure, types, naming and readme file [`#19`](https://github.com/gorhom/react-native-paper-onboarding/pull/19)
- chore: added description to onIndexChange prop [`48c552a`](https://github.com/gorhom/react-native-paper-onboarding/commit/48c552ab89e70ff70d48705201eeeeae87cc1c07)
- chore: removed unused import [`c1da684`](https://github.com/gorhom/react-native-paper-onboarding/commit/c1da684bef4bc77771bb46a23f77ac57dc7011e3)
- chore: improved circle radius calculation [`8ebaad2`](https://github.com/gorhom/react-native-paper-onboarding/commit/8ebaad260c7b5c4c8a061913ec45a78756c7a25f)

#### [v0.5.0](https://github.com/gorhom/react-native-paper-onboarding/compare/v0.4.0...v0.5.0)

> 14 May 2020

- feat: added support for a custom closeButton [`#18`](https://github.com/gorhom/react-native-paper-onboarding/pull/18)
- feat: enable content interaction (#16) [`3da8695`](https://github.com/gorhom/react-native-paper-onboarding/commit/3da8695875e6c24c41c488dab2b626012899b9e6)
- chore: updated dependencies [`e568a5b`](https://github.com/gorhom/react-native-paper-onboarding/commit/e568a5bd74b7cce2dd24a3f56f58d648153454b3)
- chore: updated examples [`7f3c9ad`](https://github.com/gorhom/react-native-paper-onboarding/commit/7f3c9ad50241f2a64d58c5e099ea37e605fcd7b3)

#### [v0.4.0](https://github.com/gorhom/react-native-paper-onboarding/compare/v0.3.1...v0.4.0)

> 21 April 2020

- feat: added support for custom view [`#14`](https://github.com/gorhom/react-native-paper-onboarding/pull/14)
- chore: update react native to v0.62.2 [`#13`](https://github.com/gorhom/react-native-paper-onboarding/pull/13)
- chore: added safe insets for examples [`0f9ac79`](https://github.com/gorhom/react-native-paper-onboarding/commit/0f9ac7935ee434b8295c88c276db4954e7ee0f36)
- chore: added github issue templates [`5f6687d`](https://github.com/gorhom/react-native-paper-onboarding/commit/5f6687dc74e8f9e616f82de4eed770dc4a1969b9)
- docs: updated readme and contributing files [`50f0618`](https://github.com/gorhom/react-native-paper-onboarding/commit/50f0618f9dccc24be53ae8821af2838eb31992b3)

#### [v0.3.1](https://github.com/gorhom/react-native-paper-onboarding/compare/v0.3.0...v0.3.1)

> 4 April 2020

- fix: updated types file path [`3dca3cc`](https://github.com/gorhom/react-native-paper-onboarding/commit/3dca3ccfeec24f3e168e8db03c3371d5b3ef0ddd)

#### [v0.3.0](https://github.com/gorhom/react-native-paper-onboarding/compare/v0.2.3...v0.3.0)

> 3 April 2020

- feat: add vertical gesture support [`#9`](https://github.com/gorhom/react-native-paper-onboarding/pull/9)
- chore: exclude example filder types from distribution [`#8`](https://github.com/gorhom/react-native-paper-onboarding/pull/8)
- chore: add @gorhom/showcase-template for example [`#7`](https://github.com/gorhom/react-native-paper-onboarding/pull/7)

#### [v0.2.3](https://github.com/gorhom/react-native-paper-onboarding/compare/v0.2.2...v0.2.3)

> 3 March 2020

- fix: moved circle animation to native thread ( by @owinter86 ) [`#6`](https://github.com/gorhom/react-native-paper-onboarding/pull/6)

#### [v0.2.2](https://github.com/gorhom/react-native-paper-onboarding/compare/v0.2.1...v0.2.2)

> 1 March 2020

- docs: update readme to include peer dependencies [`f5a44fc`](https://github.com/gorhom/react-native-paper-onboarding/commit/f5a44fc65111ae07980f58a7fd4c189f0617b43d)

#### [v0.2.1](https://github.com/gorhom/react-native-paper-onboarding/compare/v0.2.0...v0.2.1)

> 1 March 2020

- docs: add changelog [`637406b`](https://github.com/gorhom/react-native-paper-onboarding/commit/637406b24a1fed892993cacc30c219f0e84a4c64)
- chore: updated package description [`a2ad2c8`](https://github.com/gorhom/react-native-paper-onboarding/commit/a2ad2c81ba54cc4f1842dfaec61e5911fa87d617)
- chore: update package to be public [`7f27066`](https://github.com/gorhom/react-native-paper-onboarding/commit/7f2706682ec8bc0bdfbe6f70da92f9111b61ff6b)

#### v0.2.0

> 1 March 2020

- docs: update readme and add preview [`#3`](https://github.com/gorhom/react-native-paper-onboarding/pull/3)

src/types.d.ts
import type { ReactNode, MutableRefObject, FC, ComponentClass } from 'react';
import type { TextStyle, Insets, StyleProp } from 'react-native';
import type Animated from 'react-native-reanimated';

export interface PaperOnboardingMethods {
  /**
   * Navigate to next page.
   */
  next: () => void;
  /**
   * Navigate to previous page.
   */
  previous: () => void;
}

export interface PaperOnboardingItemType {
  /**
   * Image cover.
   * @type {(() => React.ReactNode) | React.ReactNode}
   */
  image?: (() => ReactNode) | ReactNode;
  /**
   * Indicator icon.
   * @type {((props: IndicatorIconProps) => React.ReactNode) | React.ReactNode}
   */
  icon?: FC<IndicatorIconProps> | ComponentClass<IndicatorIconProps>;
  /**
   * Slide/page content.
   * @type {((props: PageContentProps) => React.ReactNode) | React.ReactNode}
   */
  content?: FC<PageContentProps> | ComponentClass<PageContentProps>;
  /**
   * Title text.
   * @type {string}
   */
  title?: string;
  /**
   * Title text style.
   * @type {StyleProp<TextStyle>}
   */
  titleStyle?: StyleProp<TextStyle>;
  /**
   * Description text.
   * @type {string}
   */
  description?: string;
  /**
   * Description text style.
   * @type {StyleProp<TextStyle>}
   */
  descriptionStyle?: StyleProp<TextStyle>;
  /**
   * Background color.
   * @type {string}
   */
  backgroundColor: string;
  /**
   * Show close button when page active.
   * `note: last page will always show close button.`
   * @type {boolean}
   * @default false
   */
  showCloseButton?: boolean;
}

export interface PaperOnboardingScreenDimensions {
  width: number;
  height: number;
}

export interface PaperOnboardingIndicatorConfig {
  /**
   * Indicator size (width and height).
   * @type {number}
   * @default 40
   */
  indicatorSize: number;
  /**
   * Indicator background color.
   * @type {string}
   * @default 'white'
   */
  indicatorBackgroundColor: string;
  /**
   * Indicator border color.
   * @type {string}
   * @default 'white'
   */
  indicatorBorderColor: string;
}

export interface PaperOnboardingCloseButtonConfig {
  /**
   * Custom component to be used instead of the default close button.
   * @type {(() => React.ReactNode) | React.ReactNode}
   */
  closeButton: (() => ReactNode) | ReactNode;
  /**
   * Close button text.
   * @type {string}
   * @default close
   */
  closeButtonText: string;
  /**
   * Close button text style.
   * @type {StyleProp<TextStyle>}
   */
  closeButtonTextStyle: StyleProp<TextStyle>;
  /**
   * Callback on close button pressed.
   * @type {() => void}
   */
  onCloseButtonPress: () => void;
}

export type PaperOnboardingDirectionType = 'horizontal' | 'vertical';

// PROPS

export interface CloseButtonProps extends PaperOnboardingCloseButtonConfig {
  data: Array<PaperOnboardingItemType>;
  safeInsets: Required<Insets>;
  animatedIndex: Animated.Node<number>;
}

export interface PageProps {
  index: number;
  item: PaperOnboardingItemType;
  animatedIndex: Animated.Node<number>;
  indicatorSize: number;
  titleStyle?: StyleProp<TextStyle>;
  descriptionStyle?: StyleProp<TextStyle>;
  screenDimensions: PaperOnboardingScreenDimensions;
  safeInsets: Required<Insets>;
  handleRef: (ref: MutableRefObject<Animated.View>, index: number) => void;
}

export interface PageContentProps
  extends Pick<
    PaperOnboardingItemType,
    'title' | 'titleStyle' | 'description' | 'descriptionStyle' | 'image'
  > {
  index: number;
  animatedFocus: Animated.Node<number>;
}

export interface IndicatorIconProps {
  size: number;
}

export interface PaperOnboardingProps
  extends Pick<
      Partial<PaperOnboardingItemType>,
      'titleStyle' | 'descriptionStyle'
    >,
    Partial<PaperOnboardingIndicatorConfig>,
    Partial<PaperOnboardingCloseButtonConfig> {
  /**
   * Callback when index change.
   * @type {() => void}
   * @default noop
   */
  onIndexChange?: (index: number) => void;
  /**
   * Array of pages/slides to present.
   * @type {Array<PaperOnboardingItemType>}
   */
  data: Array<PaperOnboardingItemType>;
  /**
   * Safe area insets, usually come from `react-native-safe-area-context`.
   * @type {Insets}
   * @default {top: 50, bottom: 50, left: 50, right: 50}
   */
  safeInsets?: Partial<Insets>;
  /**
   * Pan gesture direction.
   * @type {'horizontal' | 'vertical'}
   * @default 'horizontal'
   */
  direction?: PaperOnboardingDirectionType;
}

export interface IndicatorsContainerProps
  extends PaperOnboardingIndicatorConfig {
  animatedIndex: Animated.Node<number>;
  data: PaperOnboardingItemType[];
  animatedIndicatorsContainerPosition: Animated.Node<number>;
  safeInsets: Required<Insets>;
}

export interface IndicatorProps extends PaperOnboardingIndicatorConfig {
  index: number;
  item: PaperOnboardingItemType;
  animatedIndex: Animated.Node<number>;
}

export interface BackgroundProps
  extends Pick<PaperOnboardingIndicatorConfig, 'indicatorSize'> {
  /**
   * Array of pages/slides to present.
   * @type {Array<PaperOnboardingItemType>}
   */
  data: Array<PaperOnboardingItemType>;
  animatedIndex: Animated.Node<number>;
  animatedIndicatorsContainerPosition: Animated.Node<number>;
  screenDimensions: PaperOnboardingScreenDimensions;
  safeInsets: Required<Insets>;
}

export interface BackgroundCircleProps
  extends Pick<PaperOnboardingIndicatorConfig, 'indicatorSize'> {
  index: number;
  animatedIndex: Animated.Node<number>;
  animatedIndicatorsContainerPosition: Animated.Node<number>;
  color: string;
  extendedSize: number;
  bottomPosition: number;
  screenDimensions: PaperOnboardingScreenDimensions;
}
